db.go

```diff
@@ -381,7 +381,13 @@ func (db *DB) compact() (changes bool, err error) {
 		head := &rangeHead{
 			head: db.head,
 			mint: mint,
-			maxt: maxt,
+			// We remove 1 millisecond from maxt because block
+			// intervals are half-open: [b.MinTime, b.MaxTime). But
+			// chunk intervals are closed: [c.MinTime, c.MaxTime];
+			// so in order to make sure that overlaps are evaluated
+			// consistently, we explicitly remove the last value
+			// from the block interval here.
+			maxt: maxt - 1,
 		}
 		if _, err = db.compactor.Write(db.dir, head, mint, maxt); err != nil {
```

Comment:
If we want blocks to be closed intervals *and* not overlap, the correct fix is to drop everything else in this PR and do this instead:

```diff
diff --git a/db.go b/db.go
index 28cb14f..ec6f64f 100644
--- a/db.go
+++ b/db.go
@@ -550,7 +550,7 @@ func (db *DB) reload(deleteable ...string) (err error) {
        if len(blocks) == 0 {
                return nil
        }
-       maxt := blocks[len(blocks)-1].Meta().MaxTime
+       maxt := blocks[len(blocks)-1].Meta().MaxTime + 1
 
        return errors.Wrap(db.head.Truncate(maxt), "head truncate failed")
 }
@@ -807,7 +807,7 @@ func (db *DB) Querier(mint, maxt int64) (Querier, error) {
 
 func rangeForTimestamp(t int64, width int64) (mint, maxt int64) {
        mint = (t / width) * width
-       return mint, mint + width
+       return mint, mint + width - 1
 }
 
 // Delete implements deletion of metrics. It only has atomicity guarantees on a per-block basis.
diff --git a/head.go b/head.go
index 2c7c7ec..7926cdf 100644
--- a/head.go
+++ b/head.go
@@ -1133,7 +1133,10 @@ func (s *memSeries) cut(mint int64) *memChunk {
 
        // Set upper bound on when the next chunk must be started. An earlier timestamp
        // may be chosen dynamically at a later point.
-       _, s.nextAt = rangeForTimestamp(mint, s.chunkRange)
+       _, maxt := rangeForTimestamp(mint, s.chunkRange)
+       // rangeForTimestamp() returns the maximum timestamp for the previous
+       // chunk; we want to cut right after that, so we add 1 millisecond
+       s.nextAt = maxt + 1
 
        app, err := c.chunk.Appender()
        if err != nil {
```

(At least that's what I had originally implemented and tested.)

I'm not updating this PR right away because I'm not sure there's a consensus on whether we want `[a, b]` or `[a, b)` blocks yet. @fabxc @krasi-georgiev What's your take on this?

Reply:
Yep, while I think `[a, b]` makes sense, not sure what the others think.

I'm really sorry to have you go back and forth on the changes @BenoitKnecht :( Thanks for keeping up, really appreciate it!

Reply:
No worries :) It's not an insignificant change, it's worth making sure we make the right decision.